src/ImageSharp/Formats/Pbm/PbmDecoderCore.cs

```diff
@@ -0,0 +1,199 @@
+                case '3':
+                    // Plain PPM format: 3 components per pixel, in decimal text.
+                    this.ColorType = PbmColorType.Rgb;
+                    this.Encoding = PbmEncoding.Plain;
+                    break;
+                case '4':
+                    // Binary PBM format: 1 component per pixel, 8 picels per byte.
+                    this.ColorType = PbmColorType.BlackAndWhite;
+                    this.Encoding = PbmEncoding.Binary;
+                    break;
+                case '5':
+                    // Binary PGM format: 1 components per pixel, in binary integers.
+                    this.ColorType = PbmColorType.Grayscale;
+                    this.Encoding = PbmEncoding.Binary;
+                    break;
+                case '6':
+                    // Binary PPM format: 3 components per pixel, in binary integers.
+                    this.ColorType = PbmColorType.Rgb;
+                    this.Encoding = PbmEncoding.Binary;
+                    break;
+                case '7':
+                // PAM image: sequence of images.
+                // Not implemented yet
+                default:
+                    throw new NotImplementedException("TODO");
+            }
+
+            stream.SkipWhitespaceAndComments();
+            int width = stream.ReadDecimal();
+            stream.SkipWhitespaceAndComments();
+            int height = stream.ReadDecimal();
+            stream.SkipWhitespaceAndComments();
+            if (this.ColorType != PbmColorType.BlackAndWhite)
+            {
+                this.MaxPixelValue = stream.ReadDecimal();
+                stream.SkipWhitespaceAndComments();
+            }
+            else
+            {
+                this.MaxPixelValue = 1;
+            }
+
+            this.PixelSize = new Size(width, height);
+            this.Metadata = new ImageMetadata();
+            PbmMetadata meta = this.Metadata.GetPbmMetadata();
+            meta.Encoding = this.Encoding;
+            meta.ColorType = this.ColorType;
+            meta.MaxPixelValue = this.MaxPixelValue;
+        }
+
+        private void ProcessPixels<TPixel>(BufferedReadStream stream, Buffer2D<TPixel> pixels)
+            where TPixel : unmanaged, IPixel<TPixel>
+        {
+            if (this.Encoding == PbmEncoding.Binary)
+            {
+                BinaryDecoder.Process(this.Configuration, pixels, stream, this.ColorType, this.MaxPixelValue);
+            }
+            else
+            {
+                PlainDecoder.Process(this.Configuration, pixels, stream, this.ColorType, this.MaxPixelValue);
+            }
+        }
+
+        private void ProcessUpscaling<TPixel>(Image<TPixel> image)
+            where TPixel : unmanaged, IPixel<TPixel>
+        {
+            int maxAllocationValue = (this.MaxPixelValue > 255) ? 65535 : 255;
+            float factor = maxAllocationValue / this.MaxPixelValue;
+            image.Mutate(x => x.Brightness(factor));
```

Comment:
Do we have tests that exercising this logic? Is the scaled output consistent with other common implementations?

Reply:
I have seen scaled output in xnConvert and Irfanview.

Reply:
Filed an issue to Magicks repo [here](https://github.com/ImageMagick/ImageMagick/issues/4546)

Reply:
While we wait for Magick to fix the bug, I added some Magick compatble versions. Now we have a reference test with Magick for every ColorType and Encoding combination..

Reply:
Looks like they are pushing back and nominating Netpbm as a de facto reference implementation. Feels like we can't do anything else but conform to Netpbm for maximum compatibility.

Reply:
Just to be clear, I have enabled the reference tests with Magick for all image type. To do this, I added a Magick compatible variant of the offending images. 

Turns out Magick expects a closing space after the last pixel. My code doesn't care and accepts image with and without this space at the end.

So, my code can handle both variants. To proof this, I added both variants to the RoundTripTest and others tests as well.

I propose to keep this behavior, as I consider it more user-friendly then Magick and NetPbm.

Reply:
> I propose to keep this behavior, as I consider it more user-friendly then Magick and NetPbm.

I also think this is a good idea. I tested those files with gimp, irfanview and xnview, which also dont complain decoding it.
@antonfirsov, @JimBobSquarePants  what are your opinions on this?

Reply:
Ah sorry got confused and thought we are producing non-compliant outputs, nevermind me.

Reply:
I'm with keeping the more friendly approach. It's an odd stance to have to demand the trailing whitespace on the last input. Most decoders handle missing EOF indicators well enough (Including several in _ImageMagick_) .

Reply:
Let me double check if our Encoded images load into both Magick and NetPbm, which might not be the case now. Anyway, I'll add a check for the EOF indicator in the tests.

Reply:
After adding the EOF indicator, all ColorTypes and Encodings open ciorrectly in both NetPbm and Magick.

Reply:
OK, checked Encoding all ColorTypes and Encodings and load them back into both NetPbm and Magick. All are excepted without warnings now.